File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/lisa/LisaDrive.java
package org.firstinspires.ftc.teamcode.lisa;

import android.annotation.SuppressLint;

import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.hardware.DcMotorEx;

import org.firstinspires.ftc.teamcode.common.BunyipsController;
import org.firstinspires.ftc.teamcode.common.BunyipsOpMode;

public class LisaDrive extends BunyipsController {
    final private DcMotorEx left;
    final private DcMotorEx right;
    final private ColorSensor fws;
    final private ColorSensor dws;

    private double leftPower;
    private double rightPower;

    public LisaDrive(BunyipsOpMode opMode, DcMotorEx left, DcMotorEx right, ColorSensor fws, ColorSensor dws) {
        super(opMode);

        this.left = left;
        this.right = right;
        this.fws = fws;
        this.dws = dws;
        leftPower = 0;
        rightPower = 0;
    }

    public void setPower(double left, double right) {
        leftPower = -left;
        rightPower = -right;
    }

    public void setToBrake() {
        left.setZeroPowerBehavior(DcMotorEx.ZeroPowerBehavior.BRAKE);
        right.setZeroPowerBehavior(DcMotorEx.ZeroPowerBehavior.BRAKE);
    }

    public void setToFloat() {
        left.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.FLOAT);
        right.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.FLOAT);
    }

    public void setEncoder(boolean encoder) {
        if (encoder) {
            left.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
            right.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);

            left.setMode(DcMotor.RunMode.RUN_USING_ENCODER);
            right.setMode(DcMotor.RunMode.RUN_USING_ENCODER);

        } else {
            left.setMode(DcMotor.RunMode.RUN_USING_ENCODER);
            right.setMode(DcMotor.RunMode.RUN_USING_ENCODER);
        }
    }

    @SuppressLint("DefaultLocale")
    public void update() {
        left.setPower(leftPower);
        right.setPower(rightPower);
        getOpMode().telemetry.addLine(String.format("Left Encoder: %d, Right Encoder: %d\nFWS dist: %d cm, DWS dist: %d cm",
        left.getCurrentPosition(), right.getCurrentPosition(), fws.getDistance(DistanceUnit.CM), dws.getDistance(DistanceUnit.CM)));
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/lisa/components/LisaDrive.kt
package org.firstinspires.ftc.teamcode.lisa.components

import android.annotation.SuppressLint
import com.qualcomm.robotcore.hardware.DcMotor.ZeroPowerBehavior
import com.qualcomm.robotcore.hardware.DcMotorEx
import org.firstinspires.ftc.teamcode.common.BunyipsComponent
import org.firstinspires.ftc.teamcode.common.BunyipsOpMode
import org.firstinspires.ftc.teamcode.common.Encoder
import org.firstinspires.ftc.teamcode.common.Odometer
import java.util.Locale
import kotlin.math.abs

/**
 * Drive control system for Lisa, controlling movement.
 */
class LisaDrive(
    opMode: BunyipsOpMode,
    private val leftMotor: DcMotorEx,
    private val rightMotor: DcMotorEx
) : BunyipsComponent(opMode) {
    private var leftPower = 0.0
    private var rightPower = 0.0

    // TODO: Define these values
    private val leftEncoder = Odometer(opMode, leftMotor, null, null)
    private val rightEncoder = Odometer(opMode, rightMotor, null, null)
    /*
        These might be helpful for later, defined before in the original codebase

        UltraPlanetary HD Hex Motor
        val TICKS_PER_REVOLUTION = 28.0
        val WHEEL_DIAMETER_CM = 8.5
     */

    fun setPower(left: Double, right: Double) {
        // TODO: Consider using power curves to make the robot more controllable
        leftPower = -left
        rightPower = -right
    }

    fun setToBrake() {
        leftMotor.zeroPowerBehavior = ZeroPowerBehavior.BRAKE
        rightMotor.zeroPowerBehavior = ZeroPowerBehavior.BRAKE
    }

    fun setToFloat() {
        leftMotor.zeroPowerBehavior = ZeroPowerBehavior.FLOAT
        rightMotor.zeroPowerBehavior = ZeroPowerBehavior.FLOAT
    }

    fun setEncoders() {
        leftEncoder.track()
        rightEncoder.track()
    }

    fun resetEncoders() {
        leftEncoder.reset()
        rightEncoder.reset()
    }

    fun reachedGoal(goalMM: Double, scope: Encoder.Scope = Encoder.Scope.RELATIVE): Boolean {
        return abs(leftEncoder.travelledMM(scope)) >= abs(goalMM) && abs(
            rightEncoder.travelledMM(
                scope
            )
        ) >= abs(
            goalMM
        )
    }

    fun getTravelledDist(scope: Encoder.Scope = Encoder.Scope.RELATIVE): Pair<Double, Double> {
        return Pair(leftEncoder.travelledMM(scope), rightEncoder.travelledMM(scope))
    }

    fun getEncoderValues(scope: Encoder.Scope = Encoder.Scope.RELATIVE): Pair<Double, Double> {
        return Pair(leftEncoder.position(scope), rightEncoder.position(scope))
    }

    @SuppressLint("DefaultLocale")
    fun update() {
        leftMotor.power = leftPower
        rightMotor.power = rightPower
        opMode.addTelemetry(
            String.format(
                Locale.getDefault(),
                "Drive: L: %.2f, R: %.2f",
                leftPower,
                rightPower
            )
        )
    }
}